Add a daily build and auto-cancel redundant pipelines

orf opened the pull request:
See https://github.com/django/django/pull/13852#issuecomment-770769768

This adds:
* cancelling redundant pipelines in builds (we should add this to the next set of actions we add)
* scheduled master full builds at midnight
* ~~runs the tests on pushes to master~~

felixxm commented on .github/workflows/tests.yml:
```diff
@@ -2,6 +2,10 @@ name: Tests
   pull_request:
+  push:
+  schedule:
+    # Run all tests midnight
```
```suggestion
    # Run all tests at midnight.
```

felixxm commented on .github/workflows/tests.yml:
```diff
@@ -2,6 +2,10 @@ name: Tests
 
 on:
   pull_request:
+  push:
```
IMO we don't need this. Workflow is already started on pull requests, so for manually committed changes it should be enough to run all tests once per day.

orf replied:
I think there's definitely value in running the test suite after merging a PR into into master or a release branch? It's a bit odd not to. 

We can add a specific filter to only run it on commits to master or release branches if that's better?

felixxm replied:
If we want to move all builds to GitHub actions in the future, we need to always specify a branch. Because matrices are different. In this PR I would like to move only [javascript-master](https://djangoci.com/job/javascript-master/) so both `push` and `schedule` should be run only on master.

orf replied:
Ok! Makes sense. I've added that 👍

felixxm commented on .github/workflows/cancel_runs.yml:
```diff
@@ -0,0 +1,17 @@
+      - uses: styfle/cancel-workflow-action@0.7.0
+        with:
+          workflow_id: 'tests.yml'
+          access_token: ${{ github.token }}
```
Do we need this if we remove `push` from `tests.yml`? :thinking:

orf replied:
Yep, this cancels any relevant redundant builds from any target and required a project access token to function.

felixxm reviewed (commented):
@orf Thanks :+1:

felixxm commented on .github/workflows/tests.yml:
```diff
@@ -2,6 +2,10 @@ name: Tests
+  push:
+  schedule:
+    # Run all tests at midnight.
+    - cron: "0 0 * * *"
```
Can we add some logic to run this only when we have changes in the last 24-hours? e.g.
```bash
git log --since "1 day ago" --format=oneline master | wc -l
```

orf replied:
We can do quite easily, but I really think that is over-complicating things. There isn't a benefit I can see to only running it under those conditions: the CI is free, we're not lacking in capacity or likely to run into limits at midnight.

Running it regardless of the last time a commit was made gives us more timely information about issues with any of our test dependencies.

In general I think we should keep any Actions we add simple and vanilla to begin with. We can add optimisations later if needed.

felixxm replied:
I agree that we have a commit on master almost each day but it's not the case on stable branches (which we want to move to GitHub actions in the future), e.g. on `stable/2.2.x` we can have a commit once per few months. That's why I would like to have a ready and tested solution in this trial :pray: 

> the CI is free, we're not lacking in capacity or likely to run into limits at midnight.

Yes it's free for us, but not for out planet :seedling: :world_map: If can avoid wasting energy :factory: and resource, why not.

orf replied:
Ahh right. So cronjob events only trigger on the default project branch, not on other branches.

If we wanted to trigger daily tests on those branches we'd need something a bit different, but we can do that later.

The planetary impact of spinning up a VM to run "git log" isn't _much_ different from spinning up a VM to run "npm test", but I'll add the check regardless 👌

felixxm replied:
> The planetary impact of spinning up a VM to run "git log" isn't _much_ different from spinning up a VM to run "npm test", but I'll add the check regardless ok_hand

Yeah, but it should be significant for https://djangoci.com/job/django-2.2/ :wink:

felixxm replied:
Thanks for doing this :1st_place_medal:

orf replied:
No problem! I've looked into this now - Github have specific protections against accidentally causing recursive workflow runs (see https://docs.github.com/en/actions/reference/events-that-trigger-workflows#triggering-new-workflows-using-a-personal-access-token).

I've pushed my implementation: We have a job that runs on a schedule and loops over a list of branches. If any of the branches have a commit in the last day it triggers a new build for that branch using the API. This requires an access token to be configured outside of Github Actions, but once done it should work. I'll test it on my fork first, and we can look at merging it later.

orf replied:
With this we can also trigger branch builds. I think you need push access to the repository, but it could be useful.

![Screenshot 2021-02-11 at 19 15 17](https://user-images.githubusercontent.com/1027207/107686761-87445600-6c9d-11eb-8940-53fd6f755b5b.png)

orf replied:
Done! Just tested this and it works. The access token needs to have the `Workflow` scope.

felixxm replied:
It's not in this PR anymore. What was the reason for removing `nightly_run`? :thinking:

orf replied:
Errr.... I'm not entirely sure. That might have been a mistake. I'll re-add it now 👍

orf replied:
Done, see comment below

apollo13 commented on .github/workflows/cancel_runs.yml:
```diff
@@ -0,0 +1,17 @@
+  cancel-runs:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: styfle/cancel-workflow-action@0.7.0
```
I wonder if we should require the full commit ref or vendor the actions that actually give access to the repository. I trust github actions to not do anything dangerous, but 3rd party actions are a different story imo :)

orf replied:
Yeah this is a valid concern! I'd vote for pinning to a commit hash rather than vendoring to be honest.

orf replied:
I've pinned it to the commit hash for v0.8.0

apollo13 commented on .github/workflows/nightly_run.yml:
```diff
@@ -0,0 +1,32 @@
+on:
+  schedule:
+    # Run all tests at midnight.
+    - cron: "0 0 * * *"
```
I am not sure how exact cron runs for github actions but I'd suggest running those at a more random time like 3:42. midnight seems to be a time that many people might choose, so let's help github by spreading the load?

orf replied:
I've changed it to run at 2:30am. Spreading the load is noble, but I expect Github (And Azure) scales up in preparation for this or otherwise mitigates it, so I'd rather not run it at a completely arbitrary time to be honest. But 2:30am seems good?

apollo13 commented on .github/workflows/nightly_run.yml:
```diff
@@ -0,0 +1,32 @@
+        uses: benc-uk/workflow-dispatch@v1
+        with:
+          workflow: Tests
+          token: ${{ secrets.NIGHTLY_WORKFLOW_TOKEN }}
```
the normal `GITHUB_TOKEN` supplied is not enough?

apollo13 replied:
Nevermind, this is because `GITHUB_TOKEN` cannot trigger new workflows. Sorry.

felixxm commented on .github/workflows/tests.yml:
```diff
@@ -2,6 +2,10 @@ name: Tests
 
 on:
   pull_request:
+  push:
```
Can we add the same to `linters` and specify a `master` branch?
```
  push:
    branches:
    - master
```

felixxm replied:
I pushed this to a separate PR #14046.

orf commented:
I've fixed up the PR, this should be it's final form 💪. I've removed the need for a third party library to trigger the nightly builds, as it's like 4 lines of JS and we can use the inline [@actions/script](https://github.com/actions/github-script) functionality for that.

Before merging we need a `NIGHTLY_WORKFLOW_TOKEN` secret added to the repository.

felixxm commented:
Superseded by #15501.
